agent/php_wrapper.c

```diff
@@ -8,6 +8,43 @@
 #include "php_wrapper.h"
 #include "util_logging.h"
 
+nruserfn_t* nr_php_wrap_user_function_before_after(
+    const char* name,
+    size_t namelen,
+    nrspecialfn_t before_callback,
+    nrspecialfn_t after_callback) {
+  nruserfn_t* wraprec = nr_php_add_custom_tracer_named(name, namelen TSRMLS_CC);
+
+  if (wraprec) {
+    if (after_callback) {
+      if ((NULL != wraprec->special_instrumentation)
+          && (after_callback != wraprec->special_instrumentation)) {
+        nrl_verbosedebug(
+            NRL_INSTRUMENT,
+            "%s: attempting to set special_instrumentation for %.*s, but "
+            "it is already set",
+            __func__, NRSAFELEN(namelen), NRBLANKSTR(name));
+      } else {
+        wraprec->special_instrumentation = after_callback;
+      }
+    }
+    if (before_callback) {
+      if ((NULL != wraprec->special_instrumentation)
+          && (before_callback != wraprec->special_instrumentation)) {
+        nrl_verbosedebug(NRL_INSTRUMENT,
+                         "%s: attempting to set special_instrumentation_before "
+                         "for %.*s, but "
+                         "it is already set",
+                         __func__, NRSAFELEN(namelen), NRBLANKSTR(name));
+      } else {
+        wraprec->special_instrumentation_before = before_callback;
+      }
+    }
+  }
```

Comment:
Suggestion to improve readability:
```
static bool chk_special_instrumentation_not_set(nruserfn_t* wraprec,
                                                nrspecialfn_t callback) {
  if ((NULL != wraprec->special_instrumentation)
      && (callback != wraprec->special_instrumentation)) {
    return false;
  }
  return true;
}
```

```suggestion
  if (NULL == wraprec) {
    return wraprec;
  }
  if (after_callback) {
    if (!chk_special_instrumentation_not_set(wraprec, after_callback)) {
      nrl_verbosedebug(
          NRL_INSTRUMENT,
          "%s: attempting to set special_instrumentation for %.*s, but "
          "it is already set",
          __func__, NRSAFELEN(namelen), NRBLANKSTR(name));
    } else {
      wraprec->special_instrumentation = after_callback;
    }
  }
  if (before_callback) {
    if (!chk_special_instrumentation_not_set(wraprec, before_callback)) {
      nrl_verbosedebug(NRL_INSTRUMENT,
                       "%s: attempting to set special_instrumentation_before "
                       "for %.*s, but "
                       "it is already set",
                       __func__, NRSAFELEN(namelen), NRBLANKSTR(name));
    } else {
      wraprec->special_instrumentation_before = before_callback;
    }
  }
```
abstract out special instrumentation check, unwrap nested if statement and return immediately if `NULL == wraprec`

Reply:
Thanks for pointing this out since there's a bug there (fixed in [324dbba](https://github.com/newrelic/newrelic-php-agent/pull/551/commits/324dbbaec646d5ada5c4fe79eacf30e7de8915a5)), both calls shouldn't be checking the same wraprec value. Both calls would need to check different values the function wouldn't work for both.  I think in this case it's okay since function itself is basically a check for if it is set.  It's a pretty small function, and as you mentioned, we're wanting to consider reducing function calls not add more.

Reply:
we can either inline it or turn it into a macro in that case

Reply:
inline sounds fine.  Can you add that?

Reply:
added in 136249c70c332aeac014bcccf65b3a003e5ed973